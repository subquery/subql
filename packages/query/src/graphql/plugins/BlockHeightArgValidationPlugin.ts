// Copyright 2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import {UserInputError} from 'apollo-server-express';
import type {ApolloServerPlugin} from 'apollo-server-plugin-base';
import {DocumentNode, GraphQLError, visit} from 'graphql';
import {Pool} from 'pg';

function parseBlockHeightArgs(document: DocumentNode): string[] {
  const values = [];
  visit(document, {
    Argument(node) {
      if (node.name.value === 'blockHeight' && node.value.kind === 'StringValue') {
        values.push(node.value.value);
      }
    },
  });
  return values;
}

async function fetchLastProcessedHeight(pgClient: Pool, schemaName: string): Promise<bigint | null> {
  const result = await pgClient.query(`select value from "${schemaName}"._metadata WHERE key = 'lastProcessedHeight'`);
  if (!result.rowCount) {
    return null;
  }
  return BigInt(result.rows[0].value);
}

function validateBlockHeightArgs(args: string[], lastProcessedHeight: bigint): GraphQLError | null {
  for (const arg of args) {
    if (!arg.match(/^\d+$/)) {
      return new UserInputError(`Invalid block height: ${arg}`);
    }
    const value = BigInt(arg);
    if (value > lastProcessedHeight) {
      return new UserInputError(`Block height ${arg} is larger than last processed height: ${lastProcessedHeight}`);
    }
  }
  return null;
}

// Plugin to check that any provided blockHeight arg is <= lastProcessedHeight
export function BlockHeightArgValidationPlugin({dbSchema}: {dbSchema: string}): ApolloServerPlugin {
  return {
    // eslint-disable-next-line
    requestDidStart: async () => {
      return {
        async responseForOperation({context, document}) {
          const blockHeightArgs = parseBlockHeightArgs(document);
          if (!blockHeightArgs) {
            return null;
          }
          const lastProcessedHeight = await fetchLastProcessedHeight(context.pgClient, dbSchema);
          if (!lastProcessedHeight) {
            return null;
          }
          const error = validateBlockHeightArgs(blockHeightArgs, lastProcessedHeight);
          if (!error) {
            return null;
          }
          return {errors: [error]};
        },
      };
    },
  } as ApolloServerPlugin;
}
