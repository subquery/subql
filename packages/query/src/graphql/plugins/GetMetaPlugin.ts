// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import url from 'url';
import {makeExtendSchemaPlugin, gql} from 'graphile-utils';
import fetch from 'node-fetch';
import {setAsyncInterval} from '../../utils/asyncInterval';
import {argv} from '../../yargs';
const {version: packageVersion} = require('../../../package.json');

const indexerUrl = argv('indexer') as string | undefined;

type Meta = {
  lastProcessedHeight: number;
  lastProcessedTimestamp: number;
  targetHeight: number;
  chain: string;
  specName: string;
  genesisHash: string;
  indexerHealthy: boolean;
  indexerNodeVersion: string;
  queryNodeVersion: string;
};

const metaCache = {
  queryNodeVersion: packageVersion,
} as Meta;

export const GetMetaPlugin = makeExtendSchemaPlugin((build) => {
  setAsyncInterval(async () => {
    let health;
    let meta;
    try {
      meta = await fetch(new URL(`meta`, indexerUrl));
      const result = await meta.json();
      Object.assign(metaCache, result);
    } catch (e) {
      metaCache.indexerHealthy = false;
      console.warn(`Failed to fetch indexer meta, `, e.message);
    }

    try {
      health = await fetch(new URL(`health`, indexerUrl));
      metaCache.indexerHealthy = !!health.ok;
    } catch (e) {
      metaCache.indexerHealthy = false;
      console.warn(`Failed to fetch indexer health, `, e.message);
    }
  }, 10000);

  return {
    typeDefs: gql`
      type _Meta {
        lastProcessedHeight: Int
        lastProcessedTimestamp: Date
        targetHeight: Int
        chain: String
        specName: String
        genesisHash: String
        indexerHealthy: Boolean
        indexerNodeVersion: String
        queryNodeVersion: String
      }
      extend type Query {
        _meta: _Meta
      }
    `,
    resolvers: {
      Query: {
        _meta: () => metaCache,
      },
    },
  };
});
