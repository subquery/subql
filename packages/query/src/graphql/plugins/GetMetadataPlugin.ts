// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import {URL} from 'url';
import {MetaData} from '@subql/common';
import {makeExtendSchemaPlugin, gql} from 'graphile-utils';
import fetch, {Response} from 'node-fetch';
import {setAsyncInterval} from '../../utils/asyncInterval';
import {argv} from '../../yargs';

const {version: packageVersion} = require('../../../package.json');
const indexerUrl = argv('indexer') as string | undefined;

const METADATA_TYPES = {
  lastProcessedHeight: 'number',
  lastProcessedTimestamp: 'number',
  targetHeight: 'number',
  chain: 'string',
  specName: 'string',
  genesisHash: 'string',
  indexerHealthy: 'boolean',
  indexerNodeVersion: 'string',
  queryNodeVersion: 'string',
};

type MetaType = number | string | boolean;

type MetaEntry = {key: string; value: MetaType};

const metaCache = {
  queryNodeVersion: packageVersion,
} as MetaData;

async function fetchFromApi(): Promise<void> {
  let health: Response;
  let meta: Response;

  try {
    meta = await fetch(new URL(`meta`, indexerUrl));
    const result = await meta.json();
    Object.assign(metaCache, result);
  } catch (e) {
    metaCache.indexerHealthy = false;
    console.warn(`Failed to fetch indexer meta, `, e.message);
  }

  try {
    health = await fetch(new URL(`health`, indexerUrl));
    metaCache.indexerHealthy = !!health.ok;
  } catch (e) {
    metaCache.indexerHealthy = false;
    console.warn(`Failed to fetch indexer health, `, e.message);
  }
}

async function fetchFromTable(pgClient: any, schemaName: string): Promise<MetaData> {
  const metadata = {} as MetaData;
  const keys = Object.keys(METADATA_TYPES);

  const {rows} = await pgClient.query(`select key, value from "${schemaName}"._metadata WHERE key = ANY ($1)`, [keys]);

  const dbKeyValue = rows.reduce((array: MetaEntry[], curr: MetaEntry) => {
    array[curr.key] = curr.value;
    return array;
  }, {}) as {[key: string]: MetaType};

  for (const key in METADATA_TYPES) {
    if (typeof dbKeyValue[key] === METADATA_TYPES[key]) {
      metadata[key] = dbKeyValue[key];
    }
  }

  metadata.queryNodeVersion = packageVersion;

  return metadata;
}

export const GetMetadataPlugin = makeExtendSchemaPlugin((build, options) => {
  const schemaName = options.pgSchemas;
  let metadataTableExists = false;

  const tableSearch = build.pgIntrospectionResultsByKind.attribute.find(
    (attr: {class: {name: string}}) => attr.class.name === '_metadata'
  );

  if (tableSearch !== undefined) {
    metadataTableExists = true;
  }

  if (argv(`indexer`)) {
    setAsyncInterval(fetchFromApi, 10000);
  }

  return {
    typeDefs: gql`
      type _Metadata {
        lastProcessedHeight: Int
        lastProcessedTimestamp: Date
        targetHeight: Int
        chain: String
        specName: String
        genesisHash: String
        indexerHealthy: Boolean
        indexerNodeVersion: String
        queryNodeVersion: String
      }
      extend type Query {
        _metadata: _Metadata
      }
    `,
    resolvers: {
      Query: {
        _metadata: async (_parentObject, _args, context, _info): Promise<MetaData> => {
          if (metadataTableExists) {
            const metadata = await fetchFromTable(context.pgClient, schemaName);

            if (Object.keys(metadata).length > 0) {
              return metadata;
            }
          }

          if (argv(`indexer`)) {
            return metaCache;
          }

          return;
        },
      },
    },
  };
});
