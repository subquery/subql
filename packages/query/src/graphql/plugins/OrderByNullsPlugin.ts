import { makeAddPgTableOrderByPlugin, orderByAscDesc } from 'graphile-utils';
import { Build, Plugin } from 'postgraphile';

interface IOrderByNullsPlugin extends Plugin {
  // Define the plugin interface here
}

const OrderByNullsPlugin: IOrderByNullsPlugin = makeAddPgTableOrderByPlugin(
  'public', // Replace with your schema name if different
  'MyTable', // Replace with your table name
  'MY_CUSTOM_ORDER',
  (build: Build) => {
    const { pgSql: sql } = build;
    return {
      ...orderByAscDesc(build), // Spread the existing orderings

      // Custom logic for NULLS LAST
      nullsLast: (queryBuilder) => {
        queryBuilder.orderBy(
          sql.fragment`(${queryBuilder.getTableAlias()}.my_nullable_field) IS NOT NULL`,
          true
        );
      },

      // Custom logic for NULLS FIRST
      nullsFirst: (queryBuilder) => {
        queryBuilder.orderBy(
          sql.fragment`(${queryBuilder.getTableAlias()}.my_nullable_field) IS NULL`,
          true
        );
      }
    };
  }
);

export default OrderByNullsPlugin;
