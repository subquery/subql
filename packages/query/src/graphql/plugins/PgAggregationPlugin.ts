// Copyright 2020-2022 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import AddAggregateTypesPlugin from '@graphile/pg-aggregates/dist/AddAggregateTypesPlugin';
import AddConnectionAggregatesPlugin from '@graphile/pg-aggregates/dist/AddConnectionAggregatesPlugin';
import AddConnectionGroupedAggregatesPlugin from '@graphile/pg-aggregates/dist/AddConnectionGroupedAggregatesPlugin';
import AddGroupByAggregateEnumsPlugin from '@graphile/pg-aggregates/dist/AddGroupByAggregateEnumsPlugin';
import AddGroupByAggregateEnumValuesForColumnsPlugin from '@graphile/pg-aggregates/dist/AddGroupByAggregateEnumValuesForColumnsPlugin';
import AddHavingAggregateTypesPlugin from '@graphile/pg-aggregates/dist/AddHavingAggregateTypesPlugin';
import AggregateSpecsPlugin from '@graphile/pg-aggregates/dist/AggregateSpecsPlugin';
import FilterRelationalAggregatesPlugin from '@graphile/pg-aggregates/dist/FilterRelationalAggregatesPlugin';
import InflectionPlugin from '@graphile/pg-aggregates/dist/InflectionPlugin';
import {AggregateSpec, AggregateGroupBySpec} from '@graphile/pg-aggregates/dist/interfaces';
import OrderByAggregatesPlugin from '@graphile/pg-aggregates/dist/OrderByAggregatesPlugin';

import type {Plugin} from 'graphile-build';
import {makePluginByCombiningPlugins} from 'graphile-utils';

import {argv} from '../../yargs';

const unsafe = argv('unsafe') as boolean;

// overwrite the official plugin: https://github.com/graphile/pg-aggregates/blob/main/src/AggregateSpecsPlugin.ts
// Removes all aggregation functions when not using --unsafe flag.

const AggregateSpecsPluginSafe: Plugin = (builder) => {
  builder.hook('build', (build) => {
    const pgAggregateSpecs: AggregateSpec[] = [];
    const pgAggregateGroupBySpecs: AggregateGroupBySpec[] = [];

    return build.extend(build, {
      pgAggregateSpecs,
      pgAggregateGroupBySpecs,
    });
  });
};

const plugins = [
  InflectionPlugin,
  AddGroupByAggregateEnumsPlugin,
  AddGroupByAggregateEnumValuesForColumnsPlugin,
  AddHavingAggregateTypesPlugin,
  AddAggregateTypesPlugin,
  AddConnectionAggregatesPlugin,
  AddConnectionGroupedAggregatesPlugin,
  OrderByAggregatesPlugin,
  FilterRelationalAggregatesPlugin,
];

let PgAggregationPlugin: Plugin;

if (unsafe) {
  PgAggregationPlugin = makePluginByCombiningPlugins(...plugins, AggregateSpecsPlugin);
} else {
  PgAggregationPlugin = makePluginByCombiningPlugins(...plugins, AggregateSpecsPluginSafe);
}

export default PgAggregationPlugin;
