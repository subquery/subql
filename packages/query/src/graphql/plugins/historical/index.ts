// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

import {PgBlockHeightPlugin} from './PgBlockHeightPlugin';
import PgConnectionArgFilterBackwardRelationsPlugin from './PgConnectionArgFilterBackwardRelationsPlugin';
import PgConnectionArgFilterForwardRelationsPlugin from './PgConnectionArgFilterForwardRelationsPlugin';

const historicalPlugins = [
  PgBlockHeightPlugin, // This must be before the other plugins to ensure the context is set
  PgConnectionArgFilterBackwardRelationsPlugin,
  PgConnectionArgFilterForwardRelationsPlugin,
];

export default historicalPlugins;
