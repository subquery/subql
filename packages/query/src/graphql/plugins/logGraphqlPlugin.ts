// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

import {ApolloServerPlugin} from 'apollo-server-plugin-base';
import {GraphQLRequestListener} from 'apollo-server-plugin-base/src/index';
import {GraphQLRequestContext} from 'apollo-server-types';
import {stringify} from 'flatted';

/* eslint-disable @typescript-eslint/require-await */
export const LogGraphqlPlugin: ApolloServerPlugin = {
  async requestDidStart<MyApolloContext>(
    requestContext: GraphQLRequestContext<MyApolloContext>
  ): Promise<GraphQLRequestListener<MyApolloContext>> {
    //IntrospectionQuery payload clutters logs and isn't useful
    if (requestContext.request.operationName === 'IntrospectionQuery') {
      requestContext.response.http.headers.set('message', 'Graphql IntrospectionQuery complete');
      return;
    }

    return {
      async willSendResponse({response}) {
        let message = '';

        if (!response.errors) {
          message = 'Graphql request completed successfully';
        } else {
          message = 'Encountered errors during parsing, validating, or executing the GraphQL query';
        }

        requestContext.response.http.headers.set('message', message);
      },
      async didEncounterErrors(requestContext) {
        requestContext.response.http.headers.set('stack', stringify(requestContext.errors));
      },
    };
  },
};
