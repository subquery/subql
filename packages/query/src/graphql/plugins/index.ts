// Copyright 2020-2021 OnFinality Limited authors & contributors
// SPDX-License-Identifier: Apache-2.0

/* eslint-disable */
import {
  SwallowErrorsPlugin,
  StandardTypesPlugin,
  NodePlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  ClientMutationIdDescriptionPlugin,
  MutationPayloadQueryPlugin,
  AddQueriesToSubscriptionsPlugin,
  TrimEmptyDescriptionsPlugin,
} from 'graphile-build/node8plus/plugins';
import PgBasicsPlugin from 'graphile-build-pg/node8plus/plugins/PgBasicsPlugin';
import PgIntrospectionPlugin from 'graphile-build-pg/node8plus/plugins/PgIntrospectionPlugin';
import PgTypesPlugin from 'graphile-build-pg/node8plus/plugins/PgTypesPlugin';
import PgTablesPlugin from 'graphile-build-pg/node8plus/plugins/PgTablesPlugin';
import PgConnectionArgOrderBy from 'graphile-build-pg/node8plus/plugins/PgConnectionArgOrderBy';
import PgConnectionArgOrderByDefaultValue from 'graphile-build-pg/node8plus/plugins/PgConnectionArgOrderByDefaultValue';
import PgConditionComputedColumnPlugin from 'graphile-build-pg/node8plus/plugins/PgConditionComputedColumnPlugin';
import PgAllRows from 'graphile-build-pg/node8plus/plugins/PgAllRows';
import PgColumnsPlugin from 'graphile-build-pg/node8plus/plugins/PgColumnsPlugin';
import PgColumnDeprecationPlugin from 'graphile-build-pg/node8plus/plugins/PgColumnDeprecationPlugin';
import PgForwardRelationPlugin from 'graphile-build-pg/node8plus/plugins/PgForwardRelationPlugin';
import PgRowByUniqueConstraint from 'graphile-build-pg/node8plus/plugins/PgRowByUniqueConstraint';
import PgComputedColumnsPlugin from 'graphile-build-pg/node8plus/plugins/PgComputedColumnsPlugin';
import PgQueryProceduresPlugin from 'graphile-build-pg/node8plus/plugins/PgQueryProceduresPlugin';
import PgOrderAllColumnsPlugin from 'graphile-build-pg/node8plus/plugins/PgOrderAllColumnsPlugin';
import PgOrderComputedColumnsPlugin from 'graphile-build-pg/node8plus/plugins/PgOrderComputedColumnsPlugin';
import PgOrderByPrimaryKeyPlugin from 'graphile-build-pg/node8plus/plugins/PgOrderByPrimaryKeyPlugin';
import PgRowNode from 'graphile-build-pg/node8plus/plugins/PgRowNode';
import PgNodeAliasPostGraphile from 'graphile-build-pg/node8plus/plugins/PgNodeAliasPostGraphile';
import PgRecordReturnTypesPlugin from 'graphile-build-pg/node8plus/plugins/PgRecordReturnTypesPlugin';
import PgRecordFunctionConnectionPlugin from 'graphile-build-pg/node8plus/plugins/PgRecordFunctionConnectionPlugin';
import PgScalarFunctionConnectionPlugin from 'graphile-build-pg/node8plus/plugins/PgScalarFunctionConnectionPlugin';
import PageInfoStartEndCursor from 'graphile-build-pg/node8plus/plugins/PageInfoStartEndCursor';
import PgConnectionTotalCount from 'graphile-build-pg/node8plus/plugins/PgConnectionTotalCount';

import PgSimplifyInflectorPlugin from '@graphile-contrib/pg-simplify-inflector';
import PgManyToManyPlugin from '@graphile-contrib/pg-many-to-many';
import ConnectionFilterPlugin from 'postgraphile-plugin-connection-filter';

// custom plugins
import PgConnectionArgFirstLastBeforeAfter from './PgConnectionArgFirstLastBeforeAfter';
import PgBackwardRelationPlugin from './PgBackwardRelationPlugin';
import {GetMetadataPlugin} from './GetMetadataPlugin';
import {smartTagsPlugin} from './smartTagsPlugin';
import {makeAddInflectorsPlugin} from 'graphile-utils';

/* eslint-enable */

export const defaultPlugins = [
  SwallowErrorsPlugin,
  StandardTypesPlugin,
  NodePlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  ClientMutationIdDescriptionPlugin,
  MutationPayloadQueryPlugin,
  AddQueriesToSubscriptionsPlugin,
  TrimEmptyDescriptionsPlugin,
];

export const pgDefaultPlugins = [
  PgBasicsPlugin,
  PgIntrospectionPlugin,
  PgTypesPlugin,
  // PgJWTPlugin,
  PgTablesPlugin,
  PgConnectionArgFirstLastBeforeAfter,
  PgConnectionArgOrderBy,
  PgConnectionArgOrderByDefaultValue,
  PgConditionComputedColumnPlugin,
  PgAllRows,
  PgColumnsPlugin,
  PgColumnDeprecationPlugin,
  PgForwardRelationPlugin,
  PgBackwardRelationPlugin,
  PgRowByUniqueConstraint,
  PgComputedColumnsPlugin,
  PgQueryProceduresPlugin,
  PgOrderAllColumnsPlugin,
  PgOrderComputedColumnsPlugin,
  PgOrderByPrimaryKeyPlugin,
  PgRowNode,
  PgNodeAliasPostGraphile,
  PgRecordReturnTypesPlugin,
  PgRecordFunctionConnectionPlugin,
  PgScalarFunctionConnectionPlugin, // For PostGraphile compatibility
  PageInfoStartEndCursor, // For PostGraphile compatibility
  PgConnectionTotalCount,
];

export const plugins = [
  ...defaultPlugins,
  ...pgDefaultPlugins,
  PgSimplifyInflectorPlugin,
  PgManyToManyPlugin,
  ConnectionFilterPlugin,
  smartTagsPlugin,
  GetMetadataPlugin,
  makeAddInflectorsPlugin((inflectors) => {
    const {constantCase: oldConstantCase} = inflectors;
    const enumValues = new Set();
    return {
      enumName: (v: string) => {
        enumValues.add(v);
        return v;
      },
      constantCase: (v: string) => {
        // We don't want to change the names of all enum values to CONSTANT CASE
        // because they could be specified in non CONSTANT CASE in their schema.graphql
        if (enumValues.has(v)) {
          return v;
        } else {
          return oldConstantCase(v);
        }
      },
    };
  }, true),
];
