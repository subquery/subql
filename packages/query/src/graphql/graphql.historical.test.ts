// Copyright 2020-2024 SubQuery Pte Ltd authors & contributors
// SPDX-License-Identifier: GPL-3.0

import {getPostGraphileBuilder} from '@subql/x-postgraphile-core';
import {ApolloServer, ExpressContext, gql} from 'apollo-server-express';
import {Pool} from 'pg';
import {Config} from '../configure';
import {getYargsOption} from '../yargs';
import {plugins} from './plugins';

jest.mock('../yargs', () => jest.createMockFromModule('../yargs'));

(getYargsOption as jest.Mock).mockImplementation(() => {
  return {argv: {name: 'test', aggregate: true}};
});

describe('GraphqlHistorical', () => {
  const dbSchema = 'subquery_2';

  const config = new Config({});

  const pool: Pool = new Pool({
    user: config.get('DB_USER'),
    password: config.get('DB_PASS'),
    host: config.get('DB_HOST_READ') ?? config.get('DB_HOST'),
    port: config.get('DB_PORT'),
    database: config.get('DB_DATABASE'),
  });

  pool.on('error', (err) => {
    console.error('PostgreSQL client generated error: ', err.message);
  });

  let server: ApolloServer<ExpressContext>;
  let sqlSpy: jest.SpyInstance<void, [queryText: string, values: any[], callback?: any]>;

  async function createApolloServer() {
    const builder = await getPostGraphileBuilder(pool, [dbSchema], {
      replaceAllPlugins: plugins,
      subscriptions: true,
      dynamicJson: true,
    });

    const schema = builder.buildSchema();

    const server = new ApolloServer({
      schema,
      context: {
        pgClient: pool,
      },
    });

    return server;
  }

  beforeAll(async () => {
    await pool.query(`CREATE SCHEMA IF NOT EXISTS ${dbSchema}`);
    await pool.query(`CREATE TABLE IF NOT EXISTS "${dbSchema}".listings (
      id text NOT NULL,
      item_id text NOT NULL,
      collection_id text NOT NULL,
      price_amount numeric NOT NULL,
      price_token text NULL,
      expires_at jsonb NULL,
      created_at_block_height numeric NOT NULL,
      created_at_block_time timestamp NOT NULL,
      created_at_tx_hash text NOT NULL,
      updated_at_block_height numeric NULL,
      updated_at_block_time timestamp NULL,
      updated_at_tx_hash text NULL,
      "_id" uuid NOT NULL,
      "_block_range" int8range NOT NULL,
      CONSTRAINT listings_pkey PRIMARY KEY (_id)
    );`);
    await pool.query(`CREATE TABLE IF NOT EXISTS "${dbSchema}".items (
      id text NOT NULL,
      collection_id text NOT NULL,
      token_id text NOT NULL,
      owner_id text NOT NULL,
      token_uri text NULL,
      "extension" text NULL,
      metadata jsonb NULL,
      last_traded_price_amount numeric NULL,
      approved bool NOT NULL,
      created_at_block_height numeric NOT NULL,
      created_at_block_time timestamp NOT NULL,
      created_at_tx_hash text NOT NULL,
      updated_at_block_height numeric NULL,
      updated_at_block_time timestamp NULL,
      updated_at_tx_hash text NULL,
      "_id" uuid NOT NULL,
      "_block_range" int8range NOT NULL,
      CONSTRAINT items_pkey PRIMARY KEY (_id)
    );`);
    await pool.query(`COMMENT ON TABLE "${dbSchema}".listings IS '@foreignFieldName listings
@foreignKey (item_id) REFERENCES items (id)|@singleForeignFieldName listing';`);
    await pool.query(`COMMENT ON TABLE "${dbSchema}".items IS '@foreignFieldName items';`);

    server = await createApolloServer();
    sqlSpy = jest.spyOn(pool, 'query');
  });

  beforeEach(() => {
    sqlSpy.mockClear();
  });

  afterAll(async () => {
    await pool.query(`DROP TABLE ${dbSchema}.listings;`);
    await pool.query(`DROP TABLE ${dbSchema}.items;`);
    await pool.query(`DROP SCHEMA ${dbSchema};`);
    await pool.end();
  });

  it('to filter historical items when ordering', async () => {
    const GQL_QUERY = gql`
      query nfts {
        items(orderBy: LAST_TRADED_PRICE_AMOUNT_ASC) {
          nodes {
            listings {
              nodes {
                priceAmount
              }
            }
          }
        }
      }
    `;

    const res = await server.executeOperation({query: GQL_QUERY});
    // expect(res.errors).toBeUndefined();

    expect(sqlSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('to filter historical top level', async () => {
    const GQL_QUERY = gql`
      query NFTsOnSale {
        items(filter: {listingsExist: true}) {
          nodes {
            id
            listings {
              nodes {
                id
              }
            }
          }
          totalCount
        }
      }
    `;

    const res = await server.executeOperation({query: GQL_QUERY});
    // expect(res.errors).toBeUndefined();

    expect(sqlSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('to filter historical nested (forward)', async () => {
    const GQL_QUERY = gql`
      query {
        listings(filter: {item: {approved: {equalTo: true}}}) {
          nodes {
            id
          }
        }
      }
    `;

    const res = await server.executeOperation({query: GQL_QUERY});
    // expect(res.errors).toBeUndefined();

    expect(sqlSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('to filter historical nested (backward)', async () => {
    const GQL_QUERY = gql`
      query {
        items(filter: {listings: {some: {priceToken: {equalTo: "foo"}}}}) {
          nodes {
            id
          }
        }
      }
    `;

    const res = await server.executeOperation({query: GQL_QUERY});
    // expect(res.errors).toBeUndefined();

    expect(sqlSpy.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should filter items by blockRange', async () => {
    const GQL_QUERY = gql`
      query {
        items(blockRange: [0, 100]) {
          nodes {
            id
            createdAtBlockHeight
          }
        }
      }
    `;

    const res = await server.executeOperation({query: GQL_QUERY});
    // expect(res.errors).toBeUndefined();

    expect(sqlSpy.mock.calls[0][0]).toMatchSnapshot();
  });
});
