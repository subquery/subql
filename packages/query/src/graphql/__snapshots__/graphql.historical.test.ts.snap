// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphqlHistorical to filter historical items when ordering 1`] = `
"with __local_0__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_1__."_id"), '@listings'::text, (with __local_2__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_3__."_id"), 'priceAmount'::text, ((__local_3__."price_amount"))::text))) as "@nodes" from (select 
__local_3__.*
from "subquery_2"."listings" as __local_3__

where (__local_3__."item_id" = __local_1__."id") and (__local_3__._block_range @> $1::bigint) and (__local_3__._block_range @> $1::bigint) and (TRUE) and (TRUE)
order by __local_3__."_id" ASC

) __local_3__), __local_4__ as (select json_agg(to_json(__local_2__)) as data from __local_2__) select json_build_object('data'::text, coalesce((select __local_4__.data from __local_4__), '[]'::json)) )))) as "@nodes" from (select 
__local_1__.*
from "subquery_2"."items" as __local_1__

where (__local_1__._block_range @> $1::bigint) and (TRUE) and (TRUE)
order by __local_1__."last_traded_price_amount" ASC,__local_1__."_id" ASC

) __local_1__), __local_5__ as (select json_agg(to_json(__local_0__)) as data from __local_0__) select coalesce((select __local_5__.data from __local_5__), '[]'::json) as "data" "
`;

exports[`GraphqlHistorical to filter historical nested (backward) 1`] = `
"with __local_0__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_1__."_id"), 'id'::text, (__local_1__."id")))) as "@nodes" from (select 
__local_1__.*
from "subquery_2"."items" as __local_1__

where (__local_1__._block_range @> $1::bigint) and (((exists(select 1 from "subquery_2"."listings" as __local_2__ where (__local_2__."item_id" = __local_1__."id") and (__local_2__._block_range @> $1::bigint) and (((__local_2__."price_token" = $2))) and (__local_2__._block_range @> $1::bigint))))) and (TRUE) and (TRUE)
order by __local_1__."_id" ASC

) __local_1__), __local_3__ as (select json_agg(to_json(__local_0__)) as data from __local_0__) select coalesce((select __local_3__.data from __local_3__), '[]'::json) as "data" "
`;

exports[`GraphqlHistorical to filter historical nested (forward) 1`] = `
"with __local_0__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_1__."_id"), 'id'::text, (__local_1__."id")))) as "@nodes" from (select 
__local_1__.*
from "subquery_2"."listings" as __local_1__

where (__local_1__._block_range @> $1::bigint) and ((      exists(
        select 1 from "subquery_2"."items" as __local_2__
        where (__local_1__."item_id" = __local_2__."id") and (__local_1__._block_range @> $1::bigint) and
          (((__local_2__."approved" = $2))) and (__local_2__._block_range @> $1::bigint)
      ))) and (TRUE) and (TRUE)
order by __local_1__."_id" ASC

) __local_1__), __local_3__ as (select json_agg(to_json(__local_0__)) as data from __local_0__) select coalesce((select __local_3__.data from __local_3__), '[]'::json) as "data" "
`;

exports[`GraphqlHistorical to filter historical top level 1`] = `
"with __local_0__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_1__."_id"), 'id'::text, (__local_1__."id"), '@listings'::text, (with __local_2__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_3__."_id"), 'id'::text, (__local_3__."id")))) as "@nodes" from (select 
__local_3__.*
from "subquery_2"."listings" as __local_3__

where (__local_3__."item_id" = __local_1__."id") and (__local_3__._block_range @> $1::bigint) and (__local_3__._block_range @> $1::bigint) and (TRUE) and (TRUE)
order by __local_3__."_id" ASC

) __local_3__), __local_4__ as (select json_agg(to_json(__local_2__)) as data from __local_2__) select json_build_object('data'::text, coalesce((select __local_4__.data from __local_4__), '[]'::json)) )))) as "@nodes" from (select 
__local_1__.*
from "subquery_2"."items" as __local_1__

where (__local_1__._block_range @> $1::bigint) and ((exists(select 1 from "subquery_2"."listings" as __local_5__ where (__local_5__."item_id" = __local_1__."id") and (__local_5__._block_range @> $1::bigint)))) and (TRUE) and (TRUE)
order by __local_1__."_id" ASC

) __local_1__), __local_6__ as (select json_agg(to_json(__local_0__)) as data from __local_0__) select coalesce((select __local_6__.data from __local_6__), '[]'::json) as "data", (
  select json_build_object('totalCount'::text, count(1))
  from "subquery_2"."items" as __local_1__
  where (__local_1__._block_range @> $1::bigint) and ((exists(select 1 from "subquery_2"."listings" as __local_5__ where (__local_5__."item_id" = __local_1__."id") and (__local_5__._block_range @> $1::bigint))))
) as "aggregates" "
`;

exports[`should filter items by blockRange 1`] = `
"with __local_0__ as (select to_json((json_build_object('__identifiers'::text, json_build_array(__local_1__."_id"), 'id'::text, (__local_1__."id"), 'createdAtBlockHeight'::text, ((__local_1__."created_at_block_height"))::text))) as "@nodes" from (select
__local_1__.*
from "subquery_2"."items" as __local_1__

where (__local_1__._block_range @> int8range($1, $2, '[]')) and (TRUE) and (TRUE)
order by __local_1__."_id" ASC

) __local_1__), __local_2__ as (select json_agg(to_json(__local_0__)) as data from __local_0__) select coalesce((select __local_2__.data from __local_2__), '[]'::json) as "data" "
`;

